### PR TITLE
[ConsumerRecord] Lay foundation for TableProducer + TableProducerServer

### PR DESCRIPTION
**Pull request by acco**

### CODE REVIEW

**Comment by acco** [comment]
* **#256** <a href="https://app.graphite.dev/github/pr/sequinstream/sequin/256?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @acco and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: lib/sequin/consumers_runtime/record_handler_behaviour.ex [hunk 1]
```diff
@@ -0,0 +1,19 @@
+defmodule Sequin.ConsumersRuntime.RecordHandlerBehaviour do
```
Comment (RTLS) [inline lib/sequin/consumers_runtime/record_handler_behaviour.ex:1]: perhaps TableRuntime or TableProducerRuntime etc.

File: lib/sequin/consumers_runtime/table_producer.ex [hunk 2]
```diff
@@ -0,0 +1,158 @@
+defmodule Sequin.ConsumersRuntime.TableProducer do
+  @moduledoc false
+  alias Sequin.Consumers.KeysetCursor
+  alias Sequin.Databases.PostgresDatabase.Table
+  alias Sequin.Postgres
+
+  # Cursor
+
+  def cursor(consumer_id) do
+    case Redix.command(:redix, ["GET", cursor_key(consumer_id)]) do
```
Comment (RTLS) [inline lib/sequin/consumers_runtime/table_producer.ex:10]: HGETALL

File: lib/sequin/consumers_runtime/table_producer.ex [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+        _ -> {:ok, KeysetCursor.cursor_from_result(table, sort_column_attnum, result)}
+      end
+    end
+  end
+
+  defp fetch_window_query(%Table{} = table, sort_column_attnum, nil, limit) do
+    order_by = KeysetCursor.order_by(table, sort_column_attnum)
+
+    sql = """
+    select *, row_number() over (order by #{order_by}) as row_num
```
Comment (RTLS) [inline lib/sequin/consumers_runtime/table_producer.ex:126]:
```suggestion
    select #{select_columns}, row_number() over (order by #{order_by}) as row_num
```

File: lib/sequin/consumers/record_consumer_config.ex [hunk 4]
```diff
@@ -0,0 +1,17 @@
+defmodule Sequin.Consumers.RecordConsumerConfig do
```
Comment (RTLS) [inline lib/sequin/consumers/record_consumer_config.ex:1]: Move this to PostgresDatabase.Table

File: lib/sequin/consumers/keyset_cursor.ex [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+        table.columns
+        |> Sequin.Enum.find!(fn column -> column.attnum == attnum end)
+        |> Map.fetch!(:name)
+
+      {name, value}
+    end)
+  end
+
+  def cursor_columns(table, sort_column_attnum) do
+    sort_column = Sequin.Enum.find!(table.columns, fn column -> column.attnum == sort_column_attnum end)
```
Comment (RTLS) [inline lib/sequin/consumers/keyset_cursor.ex:18]: this will get simplified as well

File: lib/sequin/consumers_runtime/table_producer_server.ex [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+        maybe_setup_allowances(state.test_pid)
+
+        with {:ok, conn} <- conn(database(state)) do
+          TableProducer.fetch_records_in_range(
+            conn,
+            table(state),
+            state.consumer.record_consumer_config.sort_column_attnum,
+            state.cursor_min,
+            state.cursor_max,
+            trunc(state.page_size * @page_size_multiplier)
```
Comment (RTLS) [inline lib/sequin/consumers_runtime/table_producer_server.ex:123]: remove trunc

File: lib/sequin/consumers_runtime/table_producer_server.ex [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+          )
+        end
+      end)
```
Comment (RTLS) [inline lib/sequin/consumers_runtime/table_producer_server.ex:126]: timeout?

File: lib/sequin/databases/postgres_database.ex [hunk 8]
```diff
@@ -149,6 +150,22 @@ defmodule Sequin.Databases.PostgresDatabase do
     end)
   end
 
+  def column_name_from_attnum(%PostgresDatabase{} = db, table_oid, attnum) do
```
Comment (RTLS) [inline lib/sequin/databases/postgres_database.ex:153]: rm?

File: lib/sequin/postgres.ex [hunk 9]
```diff
@@ -61,6 +61,17 @@ defmodule Sequin.Postgres do
...
+  @doc """
+  Accepts a SQL string with ? placeholders. Returns a SQL string with $1, $2, etc. placeholders.
+  """
+  def parameterize_sql(sql) do
+    question_mark_count = sql |> String.graphemes() |> Enum.count(&(&1 == "?"))
+
+    Enum.reduce(1..question_mark_count, sql, fn index, acc ->
+      String.replace(acc, "?", "$#{index}", global: false)
+    end)
+  end
```
Comment (RTLS) [inline lib/sequin/postgres.ex:73]: split to list, reduce, to_string

File: priv/repo/migrations/20240908020915_add_record_consumer_config_to_consumers.exs [hunk 10]
```diff
@@ -0,0 +1,21 @@
+defmodule Sequin.Repo.Migrations.AddRecordConsumerConfigToConsumers do
```
Comment (RTLS) [inline priv/repo/migrations/20240908020915_add_record_consumer_config_to_consumers.exs:1]: rm

File: test/sequin/consumer_backfill_worker_test.exs [hunk 11]
```diff
@@ -1,5 +1,5 @@
 defmodule Sequin.Streams.ConsumerBackfillWorkerTest do
```
Comment (RTLS) [inline test/sequin/consumer_backfill_worker_test.exs:1]: rm?

File: test/sequin/keyset_cursor_test.exs [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+          {"modified_at", "timestamp", 3, false},
+          {"uuid_col", "uuid", 1, true},
+          {"big_id", "bigint", 2, true}
+        ])
+
+      sort_column_attnum = Enum.find(table.columns, &(&1.name == "modified_at")).attnum
+      result = KeysetCursor.where(table, sort_column_attnum, "<")
+
+      assert result == ~s{("modified_at", "uuid_col", "big_id") < (?, ?, ?)}
+    end
```
Comment (RTLS) [inline test/sequin/keyset_cursor_test.exs:28]: what if they sort by a pk

File: test/sequin/table_producer_server_test.exs [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+        assert_records_match(messages, expected_characters)
+      end
+
+      assert_receive {:DOWN, _ref, :process, ^pid, :normal}
+    end
+  end
+
+  defp assert_records_match(messages, characters) do
+    assert_lists_equal(messages, characters, fn msg, character ->
+      assert_maps_equal(msg, Sequin.Map.stringify_keys(Map.from_struct(character)), ["id", "name"])
```
Comment (RTLS) [inline test/sequin/table_producer_server_test.exs:90]: indifferent_keys if you want

File: test/sequin/table_producer_test.exs [hunk 14]
```diff
@@ -0,0 +1,257 @@
...
+        assert Enum.any?(processed_characters, fn processed ->
+                 processed["id_integer"] == char.id_integer &&
+                   processed["id_string"] == char.id_string &&
+                   processed["id_uuid"] == char.id_uuid
+               end)
+      end)
+
+      # Verify that there are no duplicates
+      unique_processed = Enum.uniq_by(processed_characters, &{&1["id_integer"], &1["id_string"], &1["id_uuid"]})
+      assert length(unique_processed) == 6
```
Comment (RTLS) [inline test/sequin/table_producer_test.exs:227]: assert_lists_equal

File: test/sequin/table_producer_test.exs [hunk 15]
```diff
@@ -0,0 +1,257 @@
...
+    db = DatabasesFactory.insert_configured_postgres_database!(tables: [])
+    {:ok, tables} = Databases.tables(db)
+
+    characters_table = Enum.find(tables, &(&1.name == "characters"))
+    characters_multi_pk_table = Enum.find(tables, &(&1.name == "characters_multi_pk"))
+
+    column_attnums = %{
+      "characters" => map_column_attnums(characters_table),
+      "characters_multi_pk" => map_column_attnums(characters_multi_pk_table)
+    }
```
Comment (RTLS) [inline test/sequin/table_producer_test.exs:23]: less weird to use atoms

File: test/support/factory/consumers_factory.ex [hunk 16]
```diff
@@ -10,6 +10,7 @@ defmodule Sequin.Factory.ConsumersFactory do
   alias Sequin.Consumers.HttpEndpoint
   alias Sequin.Consumers.HttpPullConsumer
   alias Sequin.Consumers.HttpPushConsumer
+  alias Sequin.Consumers.RecordConsumerConfig
```
Comment (RTLS) [inline test/support/factory/consumers_factory.ex:13]: revert

File: test/support/factory/factory.ex [hunk 17]
```diff
@@ -51,6 +51,7 @@ defmodule Sequin.Factory do
   def postgres_object, do: Faker.Internet.domain_word()
   def name, do: to_string(Faker.Lorem.characters(10))
   def timestamp, do: Faker.DateTime.backward(365)
+  def naive_timestamp, do: NaiveDateTime.truncate(Faker.NaiveDateTime.backward(365), :second)
```
Comment (RTLS) [inline test/support/factory/factory.ex:54]: put back

File: test/support/unboxed_repo/migrations/20240816005458_create_test_tables.exs [hunk 18]
```diff
@@ -38,12 +44,12 @@ defmodule Sequin.Test.UnboxedRepo.Migrations.CreateTestTables do
...
       add :birth_date, :date
       add :last_seen, :time
-      add :created_at, :naive_datetime
-      add :updated_at, :utc_datetime
       add :powers, {:array, :string}
       add :metadata, :map
       add :rating, :decimal
       add :avatar, :binary
+
+      timestamps(type: :utc_datetime_usec)
```
Comment (RTLS) [inline test/support/unboxed_repo/migrations/20240816005458_create_test_tables.exs:52]: ?

**Review by RTLS — APPROVED** [review]

**Pull request merged**